### PR TITLE
refactor: adds `StorageReader` / `HashingWriter` / `HistoryWriter`

### PR DESCRIPTION
**Pull request by joshieDo**
Adds the following provider traits and moves the storage related methods into their `DatabaseProvider` implementations:
* `StorageReader`
* `HashingWriter` (will move the account methods in another PR as discussed )
* `HistoryWriter`(will move the account methods in another PR as discussed )

Also, removed the `TransactionError` which was tied to the previous `struct Transaction`. The tests inside `transaction.rs` shall be moved elsewhere in another PR.

### CODE REVIEW

File: crates/interfaces/src/provider.rs
Comment (mattsse) [inline crates/interfaces/src/provider.rs:8]: I like this, 

I'd really like to use Provider error everywhere: #2719

File: crates/storage/provider/src/traits/storage.rs
Comment (mattsse) [inline crates/storage/provider/src/traits/storage.rs:14]: I find this naming here a bit confusing.

what does `basic` mean here?

`basic` returns `plainstate` is not very intuitive to me

File: crates/storage/provider/src/transaction.rs
Comment (mattsse) [inline crates/storage/provider/src/transaction.rs:13]: is this now fully submerged into ProviderError?

**Review by mattsse — CHANGES_REQUESTED** [review]
lgtm overall.

nit re naming

File: crates/storage/provider/src/traits/storage.rs
Comment (shekhirin) [inline crates/storage/provider/src/traits/storage.rs:29]: if `with_range` means that we're passing the range, it's the case for `changed_storages` too. I think something like that may be less confusing
```suggestion
    fn changed_storages_with_blocks(
```

File: crates/storage/provider/src/traits/storage.rs
Comment (shekhirin) [inline crates/storage/provider/src/traits/storage.rs:13]: I think it should be plural and mentioning the slots, as we can retrieve multiple slots per multiple addresses
```suggestion
    /// Get plainstate storages for addresses and storage keys.
```

File: crates/storage/provider/src/transaction.rs
Comment (joshieDo) [reply crates/storage/provider/src/transaction.rs:13]: yes

File: crates/storage/provider/src/traits/storage.rs
Comment (joshieDo) [reply crates/storage/provider/src/traits/storage.rs]: added `with_range` to `changed_storages` instead to be similar to the `AccountExtReader`

File: crates/storage/provider/src/traits/storage.rs
Comment (gakonst) [inline crates/storage/provider/src/traits/storage.rs:29]: beaut

**Review by gakonst — APPROVED** [review]

**Review by shekhirin — APPROVED** [review]

**Comment by shekhirin** [comment]
pending @mattsse approval as merge is blocked until "the requested changes are addressed" 😄

**Review by mattsse — APPROVED** [review]
lgtm

**Pull request merged**
